DXGI_ERROR_INVALID_CALL on D3D12 backend Instance::create(_, _)

When creating an instance of the D3D12 backend I'm getting the following error:
```
[2019-02-16][17:31:38][gfx_backend_dx12][ERROR] Failed on dxgi factory creation: -2005270527
```
This error code is 0x887A0001 / DXGI_ERROR_INVALID_CALL and is described as
```
The application provided invalid parameter data; this must be debugged and fixed before the application is released.
```

- GFX version: latest master (c16898e327b9f007cf3a3e2c04a9afd1b3e79d0a)
- OS: Windows 10
- GPU: Nvidia GTX 1080 Ti

Okay, so even this fails:
```rust
use gfx_backend_dx12::Instance;
use std::io::prelude::*;

fn main() {
    fern::Dispatch::new()
        .format(|out, message, record| {
            out.finish(format_args!(
                "{}[{}][{}] {}",
                chrono::Local::now().format("[%Y-%m-%d %H:%M:%S]"),
                record.target(),
                record.level(),
                message
            ))
        })
        .level(log::LevelFilter::Trace)
        .chain(fern::log_file("log.txt").unwrap())
        .apply()
        .unwrap();

    #[cfg(debug_assertions)]
    log::info!("Starting in Debug Mode...");
    #[cfg(not(debug_assertions))]
    log::info!("Starting in Release Mode...");

    log::info!("Creating instance...");
    let _instance = Instance::create("Foo", 1);
    log::info!("Possibly created instance?");
    println!("Hello, world!");
    std::io::stdin().read_exact(&mut [0]).unwrap();
}
```
with
```
[2019-02-18 19:20:23][gfx_dx12_bug][INFO] Starting in Debug Mode...
[2019-02-18 19:20:23][gfx_dx12_bug][INFO] Creating instance...
[2019-02-18 19:20:23][gfx_backend_dx12][ERROR] Failed on dxgi factory creation: -2005270527
[2019-02-18 19:20:23][gfx_dx12_bug][INFO] Possibly created instance?
```
in both debug and release.

Alright, I just confirmed that this indeed fixes the bug. 
Running this:
```rust
    let mut debug_controller: *mut winapi::um::d3d12sdklayers::ID3D12Debug = std::ptr::null_mut();
    let hr = unsafe {
        winapi::um::d3d12::D3D12GetDebugInterface(
            &winapi::um::d3d12sdklayers::ID3D12Debug::uuidof(),
            &mut debug_controller as *mut *mut _ as *mut *mut _,
        )
    };

    if !winerror::SUCCEEDED(hr) {
        log::error!("Failed on getting d3d12 debug interface: {:08x?}", hr);
    } else {
        log::info!("Succeeded with getting d3d12 debug interface");
    }

    let mut queue = d3d12::WeakPtr::<dxgidebug::IDXGIInfoQueue>::null();
    let hr = unsafe {
        dxgi1_3::DXGIGetDebugInterface1(0, &dxgidebug::IDXGIInfoQueue::uuidof(), queue.mut_void())
    };

    let flags = if !winerror::SUCCEEDED(hr) {
        log::error!("Failed on getting dxgi debug interface: {:08x?}", hr);
        0
    } else {
        log::info!("Succeeded with getting dxgi debug interface");
        dxgi1_3::DXGI_CREATE_FACTORY_DEBUG
    };

    let mut dxgi_factory = d3d12::WeakPtr::<dxgi1_4::IDXGIFactory4>::null();
    let hr = unsafe {
        dxgi1_3::CreateDXGIFactory2(
            flags,
            &dxgi1_4::IDXGIFactory4::uuidof(),
            dxgi_factory.mut_void(),
        )
    };

    if !winerror::SUCCEEDED(hr) {
        log::error!("Failed on dxgi factory creation: {:08x?}", hr);
    } else {
        log::info!("Succeeded with dxgi factory creation");
    }
```
results in
```
[2019-02-18 20:15:38][gfx_dx12_bug][ERROR] Failed on getting d3d12 debug interface: 887a002d
[2019-02-18 20:15:38][gfx_dx12_bug][ERROR] Failed on getting dxgi debug interface: 887a002d
[2019-02-18 20:15:38][gfx_dx12_bug][INFO] Succeeded with dxgi factory creation
```
with the two error codes being:
DXGI_ERROR_SDK_COMPONENT_MISSING
`The operation depends on an SDK component that is missing or mismatched.`

And then conditionally not providing the debug flag makes the dxgi factory creation succeed. So this confirms my suspicion that the DirectX 12 backend hasn't been sufficiently tested without the debugging SDK components installed.

Wow, great investigation, thank you!
Sounds like a fix is an extra check during instance creation to avoid passing that debug flag? Would you want to submit a PR?

> On Feb 18, 2019, at 23:22, Christopher Serr <notifications@github.com> wrote:
> 
> Alright, I just confirmed that this indeed fixes the bug.
> Running this:
> 
>     let mut debug_controller: *mut winapi::um::d3d12sdklayers::ID3D12Debug = std::ptr::null_mut();
>     let hr = unsafe {
>         winapi::um::d3d12::D3D12GetDebugInterface(
>             &winapi::um::d3d12sdklayers::ID3D12Debug::uuidof(),
>             &mut debug_controller as *mut *mut _ as *mut *mut _,
>         )
>     };
> 
>     if !winerror::SUCCEEDED(hr) {
>         log::error!("Failed on getting d3d12 debug interface: {:08x?}", hr);
>     } else {
>         log::info!("Succeeded with getting d3d12 debug interface");
>     }
> 
>     let mut queue = d3d12::WeakPtr::<dxgidebug::IDXGIInfoQueue>::null();
>     let hr = unsafe {
>         dxgi1_3::DXGIGetDebugInterface1(0, &dxgidebug::IDXGIInfoQueue::uuidof(), queue.mut_void())
>     };
> 
>     let flags = if !winerror::SUCCEEDED(hr) {
>         log::error!("Failed on getting dxgi debug interface: {:08x?}", hr);
>         0
>     } else {
>         log::info!("Succeeded with getting dxgi debug interface");
>         dxgi1_3::DXGI_CREATE_FACTORY_DEBUG
>     };
> 
>     let mut dxgi_factory = d3d12::WeakPtr::<dxgi1_4::IDXGIFactory4>::null();
>     let hr = unsafe {
>         dxgi1_3::CreateDXGIFactory2(
>             flags,
>             &dxgi1_4::IDXGIFactory4::uuidof(),
>             dxgi_factory.mut_void(),
>         )
>     };
> 
>     if !winerror::SUCCEEDED(hr) {
>         log::error!("Failed on dxgi factory creation: {:08x?}", hr);
>     } else {
>         log::info!("Succeeded with dxgi factory creation");
>     }
> results in
> 
> [2019-02-18 20:15:38][gfx_dx12_bug][ERROR] Failed on getting d3d12 debug interface: 887a002d
> [2019-02-18 20:15:38][gfx_dx12_bug][ERROR] Failed on getting dxgi debug interface: 887a002d
> [2019-02-18 20:15:38][gfx_dx12_bug][INFO] Succeeded with dxgi factory creation
> with the two error codes being:
> DXGI_ERROR_SDK_COMPONENT_MISSING
> The operation depends on an SDK component that is missing or mismatched.
> 
> And then conditionally not providing the debug flag makes the dxgi factory creation succeed. So this confirms my suspicion that the DirectX 12 backend hasn't been sufficiently tested without the debugging SDK components installed.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

--- a/src/backend/dx12/Cargo.toml
+++ b/src/backend/dx12/Cargo.toml
@@ -26,5 +26,5 @@ d3d12 = "0.1"
 log = { version = "0.4" }
 smallvec = "0.6"
 spirv_cross = "0.12.1"
-winapi = { version = "0.3", features = ["basetsd","d3d12","d3d12sdklayers","d3d12shader","d3dcommon","d3dcompiler","dxgi1_2","dxgi1_3","dxgi1_4","dxgi1_6","dxgiformat","dxgitype","handleapi","minwindef","synchapi","unknwnbase","winbase","windef","winerror","winnt","winuser"] }
+winapi = { version = "0.3", features = ["basetsd","d3d12","d3d12sdklayers","d3d12shader","d3dcommon","d3dcompiler","dxgi1_2","dxgi1_3","dxgi1_4","dxgi1_6","dxgidebug","dxgiformat","dxgitype","handleapi","minwindef","synchapi","unknwnbase","winbase","windef","winerror","winnt","winuser"] }
 winit = { version = "0.18", optional = true }

--- a/src/backend/dx12/src/lib.rs
+++ b/src/backend/dx12/src/lib.rs
@@ -31,7 +31,7 @@ use hal::{error, format as f, image, memory, Features, Limits, QueueType, SwapIm
 
 use winapi::shared::minwindef::TRUE;
 use winapi::shared::{dxgi, dxgi1_2, dxgi1_3, dxgi1_4, dxgi1_6, winerror};
-use winapi::um::{d3d12, d3d12sdklayers, handleapi, synchapi, winbase};
+use winapi::um::{d3d12, d3d12sdklayers, dxgidebug, handleapi, synchapi, winbase};
 use winapi::Interface;
 
 use std::borrow::Borrow;
@@ -694,12 +694,33 @@ impl Instance {
             }
         }
 
+        // The `DXGI_CREATE_FACTORY_DEBUG` flag is only allowed to be passed to
+        // `CreateDXGIFactory2` if the debug interface is actually available. So
+        // we check for whether it exists first.
+        let mut queue = native::WeakPtr::<dxgidebug::IDXGIInfoQueue>::null();
+        let hr = unsafe {
+            dxgi1_3::DXGIGetDebugInterface1(
+                0,
+                &dxgidebug::IDXGIInfoQueue::uuidof(),
+                queue.mut_void(),
+            )
+        };
+
+        let factory_flags = if winerror::SUCCEEDED(hr) {
+            unsafe {
+                queue.destroy();
+            }
+            dxgi1_3::DXGI_CREATE_FACTORY_DEBUG
+        } else {
+            0
+        };
+
         // Create DXGI factory
         let mut dxgi_factory = native::WeakPtr::<dxgi1_4::IDXGIFactory4>::null();
 
         let hr = unsafe {
             dxgi1_3::CreateDXGIFactory2(
-                dxgi1_3::DXGI_CREATE_FACTORY_DEBUG,
+                factory_flags,
                 &dxgi1_4::IDXGIFactory4::uuidof(),
                 dxgi_factory.mut_void(),
             )

